storage/postgreskv/client.go

```diff
@@ -0,0 +1,280 @@
+}
+
+// GetAllPath finds all values for the provided keys, up to 100 keys
+// (in the given bucket). if more keys are provided than the maximum an
+// error will be returned.
+func (client *Client) GetAllPath(bucket storage.Key, keys storage.Keys) (storage.Values, error) {
+	if len(keys) > storage.LookupLimit {
+		return nil, storage.ErrLimitExceeded
+	}
+
+	q := `
+		SELECT metadata
+		FROM pathdata pd
+			RIGHT JOIN
+				unnest($2::BYTEA[]) WITH ORDINALITY pk(request, ord)
+			ON (pd.fullpath = pk.request AND pd.bucket = $1::BYTEA)
+		ORDER BY pk.ord
+	`
+	rows, err := client.pgConn.Query(q, []byte(bucket), pq.ByteaArray(keys.ByteSlices()))
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	values := make([]storage.Value, 0, len(keys))
+	for rows.Next() {
+		var value []byte
+		if err := rows.Scan(&value); err != nil {
+			_ = rows.Close()
+			return nil, errs.Wrap(err)
+		}
+		values = append(values, storage.Value(value))
+	}
+	_ = rows.Close()
+
+	return values, nil
+}
+
+type orderedPostgresIterator struct {
+	client         *Client
+	opts           *storage.IterateOptions
+	bucket         storage.Key
+	delimiter      byte
+	batchSize      int
+	curIndex       int
+	curRows        *sql.Rows
+	lastKeySeen    storage.Key
+	errEncountered error
+}
+
+// Next fills in info for the next item in an ongoing listing.
+func (opi *orderedPostgresIterator) Next(item *storage.ListItem) bool {
+	if !opi.curRows.Next() {
+		_ = opi.curRows.Close()
+		if opi.curIndex < opi.batchSize {
+			return false
+		}
+		if err := opi.curRows.Err(); err != nil {
+			opi.errEncountered = errs.Wrap(err)
+			return false
+		}
+		newRows, err := opi.nextQuery()
+		if err != nil {
+			opi.errEncountered = errs.Wrap(err)
+			return false
+		}
+		opi.curRows = newRows
+		opi.curIndex = 0
+		if !opi.curRows.Next() {
+			_ = opi.curRows.Close()
+			return false
+		}
+	}
+	var k, v []byte
+	err := opi.curRows.Scan(&k, &v)
+	if err != nil {
+		opi.errEncountered = errs.Wrap(err)
+		_ = opi.curRows.Close()
+		return false
+	}
+	item.Key = storage.Key(k)
+	item.Value = storage.Value(v)
+	opi.curIndex++
+	if opi.curIndex == 0 && opi.lastKeySeen.Equal(item.Key) {
+		return opi.Next(item)
+	}
+	if !opi.opts.Recurse && item.Key[len(item.Key)-1] == opi.delimiter && !item.Key.Equal(opi.opts.Prefix) {
+		item.IsPrefix = true
+		// i don't think this makes the most sense, but it's necessary to pass the storage testsuite
+		item.Value = nil
+	} else {
+		item.IsPrefix = false
+	}
+	opi.lastKeySeen = item.Key
+	return true
+}
+
+func (opi *orderedPostgresIterator) nextQuery() (*sql.Rows, error) {
+	start := opi.lastKeySeen
+	if start == nil {
+		start = opi.opts.First
+	}
+	var query string
+	if !opi.opts.Recurse {
+		if opi.opts.Reverse {
+			query = "SELECT p, m FROM list_directory_reverse($1::BYTEA, $2::BYTEA, $3::BYTEA, $4) ld(p, m)"
+		} else {
+			query = "SELECT p, m FROM list_directory($1::BYTEA, $2::BYTEA, $3::BYTEA, $4) ld(p, m)"
+		}
+	} else {
```

Comment:
The list_directory functions seem very complicated, it seems like there should be an easier way to handle it. I'm probably missing something essential, but anyways... (ignore all the postgres code bugs and +-1 errors, pseudo-code to show the ideas)

For recursive, this should LIKE queries should go directly against B-Tree so it can be implemented as:

```
SELECT fullpath, value
FROM path
WHERE fullpath LIKE {prefix}%
ORDER BY fullpath
WHERE {start} <= fullpath and fullpath <= {end} 
LIMIT $4
```

For non-recursive I'm less sure about the performance, could go either way... but the implementation could be much more concise.

```
SELECT 
    first(fullpath),
    localPath(fullpath, {prefix}, {delimiter}) as localPath,
    count(*),
    first(value)
FROM path
WHERE fullpath like {prefix}%
ORDER BY fullpath
GROUP BY localPath
WHERE ${start} <= fullpath and fullpath <= {end} 
LIMIT $4

function localPath(fullpath, prefix, delimiter) {
    // I think you have it as truncate_path, although not completely the same
    localFullPath = substring(fullpath from length(prefix))
    p = position(delimiter in localFullPath)
    if p < 0 return localFullPath
    return substring(localFullPath for p)
}
```

Reply:
For recursive queries, that's essentially what is being used already. That query is assembled in orderedPostgresIterator.nextQuery().

For non-recursive, I made a medium-large dataset of paths (all the filename paths on my workstation) and tested the performance of a variety of solutions. The performance from doing any string operations on every row, as you suggest, is prohibitively slow when there are a lot of subcomponents to skip over. Using DISTINCT on a function like this localPath() is even worse. Putting all elements into an adjacency-model tree gets barely-acceptable performance for the various types of read queries that are necessary, but is slow for inserts or moves and somewhat costly in storage space. Building trees with the nested-sets or nested-intervals models is not a great fit, because of the structure of the data. Storing materialized paths as postgresql arrays works very poorly. Postgres's own ltree type (from contrib) might work well, but it imposes limitations on the size of paths and path components that I don't think are tenable here. It would be very nice if pg could apply partitioning or DISTINCT on a LIKE prefix expression, but according to what I've been able to find, it can't. If you have conflicting data, I can investigate further.

In short, this recursive CTE is the only approach I can find which yields efficient lookups _and_ efficient inserts. As a bonus, it's nearly optimal in terms of storage space as well. The downside is the very ugly and hard-to-understand query structure.

Reply:
Ah, awesome. Can you add a relevant benchmark for the listing?

Reply:
Ok. I'll put something together.

Reply:
Some raw output from running the long benchmarks against pg, an alternate version of the pg client that doesn't use a recursive CTE, and against BoltDB:

```
~/work/storj% STORJ_POSTGRESKV_TEST='postgres://pointerdb:pg-secret-pass@localhost/pointerdb?sslmode=disable' go test -v -run xxx -bench BenchmarkSuiteLong/. ./storage/postgreskv/... -test-postgreskv-alt -test-bench-long
goos: darwin
goarch: amd64
pkg: storj.io/storj/storage/postgreskv
BenchmarkSuiteLongAlt/DeepRecursive-8         	       1	3972012305 ns/op	  12.90 MB/s
--- BENCH: BenchmarkSuiteLongAlt/DeepRecursive-8
    long_bench.go:251: Performing bulk import...
    long_bench.go:211: importing took 40.173655483s
BenchmarkSuiteLongAlt/DeepRecursiveReverse-8  	       1	3808230282 ns/op	  11.97 MB/s
BenchmarkSuiteLongAlt/DeepNonRecursive-8      	       1	9526808366 ns/op	   0.00 MB/s
BenchmarkSuiteLongAlt/DeepNonRecursiveReverse-8         	       1	9312339647 ns/op	   0.00 MB/s
BenchmarkSuiteLongAlt/ShallowRecursive-8                	       3	 460747219 ns/op	   2.68 MB/s
BenchmarkSuiteLongAlt/ShallowRecursiveReverse-8         	       3	 454055201 ns/op	   2.72 MB/s
BenchmarkSuiteLongAlt/ShallowNonRecursive-8             	      10	 151278584 ns/op	   8.17 MB/s
BenchmarkSuiteLongAlt/ShallowNonRecursiveReverse-8      	      10	 154597266 ns/op	   7.99 MB/s
BenchmarkSuiteLongAlt/TopRecursiveLimit-8               	       1	1114497280 ns/op	  91.73 MB/s
BenchmarkSuiteLongAlt/TopRecursiveLimitReverse-8        	       1	1190270426 ns/op	 106.29 MB/s
BenchmarkSuiteLongAlt/TopRecursiveStartAt-8             	       1	1100325170 ns/op	  97.58 MB/s
BenchmarkSuiteLongAlt/TopRecursiveStartAtReverse-8      	       2	 657334097 ns/op	  91.22 MB/s
BenchmarkSuiteLongAlt/TopNonRecursive-8                 	       1	38930303580 ns/op	   0.00 MB/s
BenchmarkSuiteLongAlt/TopNonRecursiveReverse-8          	       1	38593242354 ns/op	   0.00 MB/s
--- BENCH: BenchmarkSuiteLongAlt
    long_bench.go:775: Performing bulk cleanup...
    long_bench.go:233: cleanup took 95.361205ms
BenchmarkSuiteLong/DeepRecursive-8                      	       1	3932688969 ns/op	  13.03 MB/s
--- BENCH: BenchmarkSuiteLong/DeepRecursive-8
    long_bench.go:251: Performing bulk import...
    long_bench.go:211: importing took 39.139060379s
BenchmarkSuiteLong/DeepRecursiveReverse-8               	       1	3800757332 ns/op	  11.99 MB/s
BenchmarkSuiteLong/DeepNonRecursive-8                   	       1	1986910406 ns/op	   0.00 MB/s
BenchmarkSuiteLong/DeepNonRecursiveReverse-8            	       1	1298669505 ns/op	   0.00 MB/s
BenchmarkSuiteLong/ShallowRecursive-8                   	       3	 515943335 ns/op	   2.40 MB/s
BenchmarkSuiteLong/ShallowRecursiveReverse-8            	       3	 473977898 ns/op	   2.61 MB/s
BenchmarkSuiteLong/ShallowNonRecursive-8                	       1	4243869744 ns/op	   0.29 MB/s
BenchmarkSuiteLong/ShallowNonRecursiveReverse-8         	       1	2639730536 ns/op	   0.47 MB/s
BenchmarkSuiteLong/TopRecursiveLimit-8                  	       1	1074987222 ns/op	  95.10 MB/s
BenchmarkSuiteLong/TopRecursiveLimitReverse-8           	       1	1182386950 ns/op	 107.00 MB/s
BenchmarkSuiteLong/TopRecursiveStartAt-8                	       1	1111324959 ns/op	  96.61 MB/s
BenchmarkSuiteLong/TopRecursiveStartAtReverse-8         	       2	 665344640 ns/op	  90.12 MB/s
BenchmarkSuiteLong/TopNonRecursive-8                    	       2	 738233795 ns/op	   0.00 MB/s
BenchmarkSuiteLong/TopNonRecursiveReverse-8             	       2	 734726361 ns/op	   0.00 MB/s
--- BENCH: BenchmarkSuiteLong
    long_bench.go:775: Performing bulk cleanup...
    long_bench.go:233: cleanup took 93.600617ms
PASS
ok  	storj.io/storj/storage/postgreskv	231.662s
?   	storj.io/storj/storage/postgreskv/schema	[no test files]
~/work/storj% go test -v -run xxx -bench BenchmarkSuiteLong/. ./storage/boltdb/... -test-bench-long
goos: darwin
goarch: amd64
pkg: storj.io/storj/storage/boltdb
BenchmarkSuiteLong/DeepRecursive-8         	      10	 131461868 ns/op	 389.89 MB/s
--- BENCH: BenchmarkSuiteLong/DeepRecursive-8
    long_bench.go:251: Performing bulk import...
    long_bench.go:211: importing took 2m6.711399139s
BenchmarkSuiteLong/DeepRecursiveReverse-8  	      10	 142506496 ns/op	 319.80 MB/s
BenchmarkSuiteLong/DeepNonRecursive-8      	   30000	     56642 ns/op	  74.38 MB/s
BenchmarkSuiteLong/DeepNonRecursiveReverse-8         	   30000	     54868 ns/op	  76.78 MB/s
BenchmarkSuiteLong/ShallowRecursive-8                	     300	   4765679 ns/op	 259.34 MB/s
BenchmarkSuiteLong/ShallowRecursiveReverse-8         	     300	   4923633 ns/op	 251.02 MB/s
BenchmarkSuiteLong/ShallowNonRecursive-8             	     300	   5716027 ns/op	 216.23 MB/s
BenchmarkSuiteLong/ShallowNonRecursiveReverse-8      	     200	   5973946 ns/op	 206.89 MB/s
BenchmarkSuiteLong/TopRecursiveLimit-8               	       5	 301675698 ns/op	 338.88 MB/s
BenchmarkSuiteLong/TopRecursiveLimitReverse-8        	       3	 355463711 ns/op	 355.92 MB/s
BenchmarkSuiteLong/TopRecursiveStartAt-8             	       5	 304097570 ns/op	 353.08 MB/s
BenchmarkSuiteLong/TopRecursiveStartAtReverse-8      	       5	 202944901 ns/op	 295.46 MB/s
BenchmarkSuiteLong/TopNonRecursive-8                 	  100000	     13864 ns/op	  17.24 MB/s
BenchmarkSuiteLong/TopNonRecursiveReverse-8          	  100000	     13232 ns/op	  18.06 MB/s
--- BENCH: BenchmarkSuiteLong
    long_bench.go:775: Performing bulk cleanup...
    long_bench.go:233: cleanup took 258.015µs
PASS
ok  	storj.io/storj/storage/boltdb	154.925s
```

The above numbers, graphed:

<img width="1033" alt="chart of storj long_bench times" src="https://user-images.githubusercontent.com/47221/47335304-37ff3500-d650-11e8-8d97-ea65ce11f530.png">

BoltDB does extremely well for all the reading tasks, but quite poorly under heavy write load (and that is with syncs turned off). If I remember how lmdb/bolt are implemented, I'd expect bolt to start showing worse numbers in comparison to postgresql when there are more simultaneous queries. Or I might be totally wrong about that.

The alternate/straightforward PG approach does well, except in the case where there are lots of elements that need to be skipped over, such as when getting a listing of the toplevel directory. In that case, though, performance is utterly abysmal.

The recursive-CTE PG approach does more poorly when getting a listing of a directory with thousands of entries but little to no subdirectory items that need to be skipped over: a "shallow" directory. This is to be expected, as it requires extra calculation and new index lookups after every element it finds. But even in this worst case, performance is maaaaybe within the bounds of sanity.

Perhaps we could come up with a solution where both approaches are used according to some heuristics that try to predict which will do better. Or, if there are cpu cycles to spare, just run both queries each time and use whichever one returns first.

I'd like to implement some of the other solutions I mentioned, to the point that they pass tests, and get benchmark information from all of them, but I'm afraid I've probably spent more time on this than it deserves for now.